### ISSUE TITLE
Image size requirements for OpenCL BatchNorm Layer

### ISSUE THREAD

**Issue by gplhegde** [ARM-software/ComputeLibrary#202]
Hi,
       Are there any constraints on the input tensor dimensions for OpenCL Batch Norm layer? I see that the outputs are correct only when the input image width (dim 0) is multiple of 4.  If this is the requirement, what is the workaround for other image widths that are not multiple of 4? Padding? Normally the convolution output tensor is given as input to batch norm layer and thus it can be of any size. How to adding padding in such cases?

Thanks,
Gopal

**Comment by GeorgeARM** [comment]
Hello @gplhegde,

It should work for every dimension size. If you see incorrect results for non-multiples of 4 then it probably means that you allocate the input tensor before the configure stage. This will reduce the execution window if the tensor width is not a multiple of 4 (in case of floats). You should first configure your kernels (as it might add padding in case of non-multiple of 4) and then allocate the tensors.
Hope this helps.

**Comment by gplhegde** [comment]
I see the problem. Got it working by allocating the input tensor after configuring the layer.

**Comment by gplhegde** [comment]
I have another question regarding reusing the buffers among layers. For example, I have a convolution --> batch norm --> activation back to back. As batch norm and activation layers can use in-place computations,   I want to share the output tensor of the convolution layer as the input tensor for both batch norm and activation layer.
```
CLConvolutionLayer conv;
CLBatchNormalizationLayer norm;
CLActivationLayer act;
CLTensor input, weight, bias, output, mean, var, gamma, beta;

// init tensors

// configure layers
conv.configure(&input, &weight, &bias, &output);
norm.configure(&output, &output, &mean, &var, &beta, &gamma, 1e-8);
act.configure(&output, &input, ActivationLayerInfo::ActivationFunction::RELU)

// allocate tensors

```
1) In such cases, how the internal tensor padding is taken care? Because different layers may have different alignment requirements.
2) Is there any way to access this padding information? This will be useful if anyone directly want map the buffer as float * and access the elements as opposed to using buffer.ptr_to_element()

Thanks in advance!

**Comment by GeorgeARM** [comment]
Hello @gplhegde,

Unfortunately, the current revision of BatchNormalization and ActivationLayer does not support in-place computation.
Yes you can access the padding requirements of a tensor from its TensorInfo.
The **padding()** call on a TensorInfo will return the padding requirements and the **has_padding()** call will inform you about the presence of padding.
Hope this helps.

**Comment by gplhegde** [comment]
Thanks for the clarification. Hope to see support for these features soon.
